### ISSUE TITLE
Event hub sometimes stops delivering messages without error

### ISSUE THREAD

**Issue by SilverBkk** [Azure/azure-sdk-for-net#11557]
**Describe the bug**
We noticed that after some time the app simply stops reading messages (even though devices are still connected and sending messages and we can still send messages to devices). In the logs of the WebApp BackgroundWorker, we can see that after a while, some partitions stop receiving messages without error but some partitions continue receiving. Restarting the app causes messages to be read again. We follow the code sample to implement a BackgroundWorker: [Azure-Samples/azure-iot-samples-csharp/iot-hub/Quickstarts/read-d2c-messages/](https://github.com/Azure-Samples/azure-iot-samples-csharp/tree/master/iot-hub/Quickstarts/read-d2c-messages)

**Environment**
- Microsoft Azure Cloud (West Europe)
- Azure IoT Hub resource (Tier S1 Standard - IoT Devices: 93)
- Microsoft.Azure.EventHubs v4.2.0
- Windows 10, .NET Core v2.1 (netcoreapp2.1)
- Visual Studio 16.5.4

**D2C Reader BackgroundWorker**
```
public override async void Start()
{
	Logger.Info($"Start D2CMessageReaderWorker: {DateTime.UtcNow:yyyy-MM-dd HH:mm:ss}");
	try
	{
		this.eventHubsCompatibleEndpoint = settingManager.GetSettingValue(AppSettings.IotHubManagement.IotHubCompatibleEndpoint);
		this.eventHubsCompatiblePath = settingManager.GetSettingValue(AppSettings.IotHubManagement.IotHubCompatiblePath);
		this.iotHubSasKey = settingManager.GetSettingValue(AppSettings.IotHubManagement.IotHubSasKey);
		this.iotHubSasKeyName = settingManager.GetSettingValue(AppSettings.IotHubManagement.IotHubSasKeyName);
	}
	catch
	{
		this.Logger.Error("The D2CMessageReaderWorker couldn't be started because parameters are missing");
	}
	if (!string.IsNullOrEmpty(this.eventHubsCompatibleEndpoint) && !string.IsNullOrEmpty(this.eventHubsCompatiblePath) && !string.IsNullOrEmpty(this.iotHubSasKey) && !string.IsNullOrEmpty(this.iotHubSasKeyName))
	{
		// Create an EventHubClient instance to connect to the IoT Hub Event Hubs - compatible endpoint.
		var connectionString = new EventHubsConnectionStringBuilder(new Uri(eventHubsCompatibleEndpoint), eventHubsCompatiblePath, iotHubSasKeyName, iotHubSasKey);
		eventHubClient = EventHubClient.CreateFromConnectionString(connectionString.ToString());
		// Create a PartitionReciever for each partition on the hub.
		var runtimeInfo = await eventHubClient.GetRuntimeInformationAsync();
		tasks.Clear();
		foreach (string partition in runtimeInfo.PartitionIds)
		{
			tasks.Add(ReceiveMessagesFromDeviceAsync(partition, cancelToken.Token));
		}
	}else{this.Stop();}
}
private async Task ReceiveMessagesFromDeviceAsync(string partition, CancellationToken token)
{
	var eventHubReceiver = eventHubClient.CreateReceiver("$Default", partition, EventPosition.FromEnqueuedTime(DateTime.Now));
	while (true)
	{
		if (token.IsCancellationRequested){break;}
		var events = await eventHubReceiver.ReceiveAsync(20);
		if (events == null){continue;}
		foreach (EventData eventData in events)
		{
			string data = Encoding.UTF8.GetString(eventData.Body.Array);
			Logger.Info($"Message received on partition {partition}");
			Logger.Info($"  {data}");
			await this.messageManager.HandleReceivedMessage(data);
		}
	}
}	
public override void Stop()
{
	cancelToken.Cancel();
}
public override void WaitToStop()
{
	Task.WaitAll(tasks.ToArray());
}
```

**Comment by serkantkaraca** [comment]
Are you logging from both errors handlers?

[EventProcessorOptions.SetExceptionHandler](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.eventhubs.processor.eventprocessoroptions.setexceptionhandler?view=azure-dotnet-legacy#Microsoft_Azure_EventHubs_Processor_EventProcessorOptions_SetExceptionHandler_System_Action_Microsoft_Azure_EventHubs_Processor_ExceptionReceivedEventArgs__)

[IEventProcessor.ProcessError](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.eventhubs.processor.ieventprocessor.processerrorasync?view=azure-dotnet-legacy#Microsoft_Azure_EventHubs_Processor_IEventProcessor_ProcessErrorAsync_Microsoft_Azure_EventHubs_Processor_PartitionContext_System_Exception_)

**Comment by SilverBkk** [comment]
Ok, I will implement those exceptions errors to log it and I will let you know. Thanks.

**Comment by serkantkaraca** [comment]
Please ignore my initial comment. I initially thought you were implementing processor host however just realized from the code snippet that PartiitonReceiver is actually used. Can you try logging null return from ReceiveAsync call in below?

`		if (events == null){continue;}
		foreach (EventData eventData in events)
		{
			string data = Encoding.UTF8.GetString(eventData.Body.Array);
			Logger.Info($"Message received on partition {partition}");
			Logger.Info($"  {data}");
			await this.messageManager.HandleReceivedMessage(data);
		}`

**Comment by SilverBkk** [comment]
After struggling with IEventProcessor.ProcessErrorAsync (i've learned a lot of things), I've decided to implement all Errors and Exceptions in my current code. I've implemented a TRY CATCH in the infinite loop, also a IPartitionReceiveHandler to handle ProcessErrorAsync, and log everything as well as the CancellationRequested and the events == null.
The new BackgroundWorker is currently running on production server. 
I will analyze all the logs asap to understand and locate the error.
I will update my post next week to let you know. 
Thanks for your support.
